src/util/CompositeAsyncHandler.ts

```diff
@@ -0,0 +1,42 @@
+import { AsyncHandler } from './AsyncHandler';
+import { UnsupportedHttpError } from './errors/UnsupportedHttpError';
+
+export class CompositeAsyncHandler<TIn, TOut> implements AsyncHandler<TIn, TOut> {
+  private readonly handlers: AsyncHandler<TIn, TOut>[];
+
+  public constructor (handlers: AsyncHandler<TIn, TOut>[]) {
+    this.handlers = handlers;
+  }
+
+  public async canHandle (input: TIn): Promise<void> {
+    const errors: Error[] = [];
+
+    for (const handler of this.handlers) {
+      try {
+        await handler.canHandle(input);
+
+        return;
+      } catch (error) {
+        errors.push(error);
+      }
+    }
+
+    throw new UnsupportedHttpError(errors.map((error: Error): string => error.message).join(' '));
+  }
+
+  public async handle (input: TIn): Promise<TOut> {
+    const errors: Error[] = [];
+
+    for (const handler of this.handlers) {
+      try {
+        await handler.canHandle(input);
+
+        return handler.handle(input);
+      } catch (error) {
+        errors.push(error);
+      }
+    }
+
+    throw new UnsupportedHttpError(errors.map((error: Error): string => error.message).join(' '));
```

Comment:
We probably don't want to collect the `canHandle` errors here.
I would put the try-catch only over `canHandle`, and only use the error to know if this handler should be used.

If none of the handlers can handle the input, I would throw a generic error such as `All composed handlers failed`. I would add a comment to say that this situation *should* not be able to occur, only if the consumer did not call `canHandle`. We could optionally also include this in the error message for improved devx.

Reply:
Yea theoretically this one would never be thrown if everything gets used correctly. Actually makes more sense to throw the error of the handler that is running if it's an error there. I'll change that.